engines/bops_core/app/controllers/concerns/bops_core/application_controller.rb

```diff
@@ -64,12 +64,26 @@ def current_api_user
     end
 
     def authenticate_api_user!
-      unless current_api_user
+      unless current_api_user && valid_api_key_scope?(current_api_user)
         json = {error: {code: 401, message: "Unauthorized"}}
         render json: json, status: :unauthorized
       end
     end
 
+    def valid_api_key_scope?(key)
+      scope_type = if request.method == "GET"
+        ":read"
+      elsif params[:controller].include?("validation_request")
+        ":update"
+      else
+        ":submit"
+      end
+
+      raise NotImplementedError, "No permissions have been defined for this authenticated method for #{params[:controller]}:#{request.method}" if required_api_key_scope.nil?
+
+      required_api_key_scope == :any || key.permissions.include?(required_api_key_scope + scope_type)
+    end
```

Comment:
I'm not sure if there's a better approach for this.

I had initially thought of just defining a permission for each action somehow (so mapping `CommentsController#create` to `comment:submit` and so on), but because the access control is done before the action, I don't know if that's possible. So instead this is defining the first half of the scope on the controller level, which should be fine as long as we never want some actions that need `planning_application:...` scope and some that need `comment:...` scope in the same controller; then the second half of the scope comes from whether it's a GET request (read-only) or anything else (needs write permissions).